examples/pytorch/rgat/train.py

```diff
@@ -0,0 +1,123 @@
+import tqdm
+
+class HeteroGAT(nn.Module):
+    def __init__(self, etypes, in_size, hid_size, out_size, n_heads=4):
+        super().__init__()
+        self.layers = nn.ModuleList()
+        self.layers.append(dglnn.HeteroGraphConv({
+            etype: dglnn.GATConv(in_size, hid_size // n_heads, n_heads)
+            for etype in etypes}))
+        self.layers.append(dglnn.HeteroGraphConv({
+            etype: dglnn.GATConv(hid_size, hid_size // n_heads, n_heads)
+            for etype in etypes}))
+        self.layers.append(dglnn.HeteroGraphConv({
+            etype: dglnn.GATConv(hid_size, hid_size // n_heads, n_heads)
+            for etype in etypes}))
+        self.dropout = nn.Dropout(0.5)
+        self.linear = nn.Linear(hid_size, out_size)   # Should be HeteroLinear
+
+    def forward(self, blocks, x):
+        h = x
+        for l, (layer, block) in enumerate(zip(self.layers, blocks)):
+            h = layer(block, h)
+            # One thing is that h might return tensors with zero rows if the number of dst nodes
+            # of one node type is 0.  x.view(x.shape[0], -1) wouldn't work in this case.
+            h = apply_each(h, lambda x: x.view(x.shape[0], x.shape[1] * x.shape[2]))
+            if l != len(self.layers) - 1:
+                h = apply_each(h, F.relu)
+                h = apply_each(h, self.dropout)
+        return self.linear(h['paper'])
+
+def evaluate(model, dataloader, desc):
+    preds = []
+    labels = []
+    with torch.no_grad():
+        for input_nodes, output_nodes, blocks in tqdm.tqdm(dataloader, desc=desc):
+            x = blocks[0].srcdata['feat']
+            y = blocks[-1].dstdata['label']['paper'][:, 0]
+            y_hat = model(blocks, x)
+            preds.append(y_hat.cpu())
+            labels.append(y.cpu())
+        preds = torch.cat(preds, 0)
+        labels = torch.cat(labels, 0)
+        acc = MF.accuracy(preds, labels)
+        return acc
+
+def train(train_loader, val_loader, test_loader, model):
+    # loss function and optimizer
+    loss_fcn = nn.CrossEntropyLoss()
+    opt = torch.optim.Adam(model.parameters(), lr=0.001, weight_decay=5e-4)
+
+    # training loop
+    for epoch in range(10):
+        model.train()
+        total_loss = 0
+        for it, (input_nodes, output_nodes, blocks) in enumerate(tqdm.tqdm(train_dataloader, desc="Train")):
+            x = blocks[0].srcdata['feat']
+            y = blocks[-1].dstdata['label']['paper'][:, 0]
+            y_hat = model(blocks, x)
+            loss = loss_fcn(y_hat, y)
+            opt.zero_grad()
+            loss.backward()
+            opt.step()
+            total_loss += loss.item()
+        model.eval()
+        val_acc = evaluate(model, val_dataloader, 'Val. ')
+        test_acc = evaluate(model, test_dataloader, 'Test ')
+        print("Epoch {:05d} | Loss {:.4f} | Validation Acc. {:.4f} | Test Acc. {:.4f} "
+              . format(epoch, total_loss / (it + 1), val_acc.item(), test_acc.item()))
```

Comment:
We now recommend using f-strings like [this](https://github.com/dmlc/dgl/blob/f846d902cf6c5cbcd988eccb3409955415b988f0/python/dgl/nn/pytorch/network_emb.py#L293).

Reply:
Thanks. I updated accordingly.